design/20220722-publicbundle.md

```diff
@@ -0,0 +1,289 @@
+	"/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem", // CentOS/RHEL 7
+	"/etc/ssl/cert.pem",                                 // Alpine Linux
+}
+```
+
+It's a chore to manually specify a location for a trust store, and it's unfamiliar and error-prone
+for many users. However, if a user instead overwrites any of these files with their own
+`trust`-generated certificate trust bundle they'll currently overwrite their public trust bundle
+provided by their OS / container base image.
+
+Unless the user's application only talks to services which the user also controls, what users are
+likely to really want in most cases is a bundle containing both the publicly trusted certs _and_
+the user's own roots.
+
+It's possible in most distros to append to the cert bundle, but that usually involves running some
+command to generate the final bundle, which would change how the user invokes their application
+inside their container.
+
+Ideally, the whole bundle a user requires would simply be available at one of the above locations.
+
+### Goals
+
+- A user configuring a `Bundle` can add a public trust bundle at least as easily as they can add certs from a `ConfigMap` or `Secret`
+- It's clear to users exactly what version of the public bundle is used at any given time, for auditing purposes
+- A user whose application must use both public and private CAs can confidently overwrite their "system" trust store in a container using a trust `Bundle`
+- A user can update any bundle in use with any version of the trust controller and be confident that it will work
+
+### Non-Goals
+
+- Editing/pruning the public trust bundle (out of scope for the initial feature but might be desirable later)
+- Supporting [OS-specific](https://superuser.com/questions/437330/how-do-you-add-a-certificate-authority-ca-to-ubuntu) certificate update mechanisms
+- Supporting a plethora of options for sources of public trust bundles; we might want this in the future but we should start with one
+
+## Proposal
+
+We propose that the cert-manager project becomes a distributor of its own public trust bundle, which at
+least to start will be a carbon copy of the Mozilla trust bundle.
+
+That bundle in will in turn be packaged into a container which will be used as a sidecar in the trust
+controller's pod, with the ability to present its own bundle to the controller container.
+
+A deployed instance of trust will therefore contain two containers in a pod; the trust controller itself, and a bundle
+container with which trust will communicate in order to fetch a trust bundle.
+
+We'll also add a `publicBundle` field as a source, which is the method by which users include the bundle from the bundle container.
+
+```yaml
+apiVersion: trust.cert-manager.io/v1alpha1
+kind: Bundle
+metadata:
+  name: my-org.com
+spec:
+  sources:
+  - configMap:
+      name: "my-root-cm"
+      key: "root-cert.pem"
+  - publicBundle:
+      source: "cert-manager" # i.e. use the cert-manager bundle
```

Comment:
```suggestion
      name: "cert-manager" # i.e. use the cert-manager bundle
```

`source` is a stutter from the `sources` block.

Reply:
> We could also optionally allow users to select the version of this named bundle to use. The default would be the named bundle with the largest version according to SEMVAR.

I think this adds significant complications. The idea presented in this design is that to update the bundle version the entire bundle container would be changed out.

Reply:
We should still be able to select which version through the Bundle resourced to allow roll backs or pinning.

Reply:
I _really_ don't think we should allow either rollbacks or pinning in a `Bundle`. A roll back should be the person replacing the bundle container with an old version.

If I specify this:

```yaml
apiVersion: trust.cert-manager.io/v1alpha1
kind: Bundle
metadata:
  ...
spec:
  sources:
  - packagedBundle:
      name: "cert-manager"
      version: "2"
  target: ...
```

And version 2 doesn't exist (e.g. we have a static bundle which is at version "3"), what can we meaningfully do except block creation of the target?

For a static bundle the only meaningful operation is to consume the bundle as-is (because there's only one bundle ever).

For a dynamic bundle the only thing I'd want is to consume the latest (because if the user doesn't want that behaviour, why opt into receiving updates in the first place? just use a static bundle and update by hand, surely?)

Reply:
Yes, we should be blocking a sync of the target when the source doesn’t exist.

Having more than one bundle is important for users who wish to  propagate a trust bundle and don’t want the sources to be accessible via the ConfigMap or Secrets API (and therefore editable).

I never just want to consume the latest- one can imagine scenarios such as CA rotation. A user can safely upgrade all bundle side cars, and then programmatically control the rollout via the API (monitoring the target syncing by acting on the Bundle statuses). A bad rollout can be rolled back in a controlled way by changing the Bundle spec, and not the sidecar container.


There is also the case of multi-tenanted scenarios where teams want control of their destiny- version pinning solves this.

Reply:
> Yes, we should be blocking a sync of the target when the source doesn’t exist.
> Having more than one bundle is important for users who wish to propagate a trust bundle and don’t want the sources to be accessible via the ConfigMap or Secrets API (and therefore editable).

Yeah I'm convinced by supporting the addition of user-added sidecars down the road as additional sources and I agree about the blocking.

I think the initial version of the implementation of this design should take care to _enable_ the use of additional sidecars but we should keep a focus on providing a first-party public trust bundle sidecar first.

> I never just want to...

As far as I understand it, everything you've described is already solved by the proposal as written without the need for a bundle to support providing multiple versions within itself.

The whole reason the bundles are in a separate container in the first place is so they can be individually versioned separately to the version of the controller.

If I want to do a CA rotation these are the steps with the current design:

1. I start with bundle sidecar v1 with just `oldRoot`
2. I roll out bundle sidecar v2 with both `oldRoot` and `newRoot`; my target configmap is updated automatically
3. I redeploy every pod to ensure it picked up the change
4. I roll out bundle sidecar v3 with just `newRoot`; my target configmap is updated automatically
5. I redeploy every pod to ensure it picked up the change

The hard part at each step is to ensure that every pod is consuming the latest bundle. If there's a problem with sidecar v2 or v3, I roll back to any older version and redeploy every pod.

Adding versions inside the container presumably looks like this (?):

1. I start with bundle sidecar v1 which contains bundle version `A`. All `Bundle` resources set `A` as their target version [1]
2. I roll out bundle sidecar v2 which contains bundle versions `A`, `B` and `C`. Nothing changes yet.
3. I change all my `Bundle` resources to set their target version as `B`. [2]
4. I redeploy every pod to make sure it picked up the change
5. I change all my `Bundle` resources to set their target version as `C`.
6. I redeploy every pod to make sure it picked up the change
7. I roll out bundle sidecar v3 which contains bundle version `C` only
8. I redeploy every pod to make sure it picked up the change

If I detect a problem before step 7 I can roll back by changing all my bundles to point to `A` again, sure, and that's maybe a little easier than rolling back the sidecar container version in some scenarios, but I don't think that gain is worth the huge amount of added complexity and additional failure modes.

> There is also the case of multi-tenanted scenarios where teams want control of their destiny- version pinning solves this.

I don't see any issues here which couldn't be solved by the cluster admin of a multi-tenanted cluster adding additional new sidecars. Would need to see specific examples of a situation where this could go wrong and why version pinning would be required as a solution.

[1] I don't know how we make the fact that it's called `A` discoverable, either; I suspect it ends up being some doc that people have to consult which is outside of k8s, or else they'll need to check the logs of the trust controller or something

[2] There's a huge risk here that Bundles go directly from `A` -> `C` and cause an outage for themselves. That can be gotten around by making sidecar v2 just contain `A` and `B` and then v3 contain `B` and `C`, and then v4 contain just `C`, but that's really quite a brutal upgrade path and it complicates rollbacks too.